### ISSUE TITLE
BlockSharedMemStMember default min alignment bad for performance

### ISSUE THREAD

**Issue by jkelling** [alpaka-group/alpaka#1278]
The default minimum alignment for shared variables is currently set to `core::vectorization::defaultAlignment (= 32)`:

https://github.com/alpaka-group/alpaka/blob/8eea24392234dc5a913f125b531ea6545c161558/include/alpaka/block/shared/st/BlockSharedMemStMember.hpp#L28

On second thought this does not much sense even on CPU: If small variables, like `int32_t` are allocated, these will be padded out in 32-byte chunks. This has the following effects in CPU backends (e.g. `CpuSerial`, `Omp2Blocks`) in increasing order of severity, regarding this issue:
* waste of very limited memory space (size is limited for GPU portability)
* no vector instruction will actually be created for single ints, so no benefit
* neighboring small variables to not shared a cache line, reducing cache hits

The last point is reversed in backends like `Omp2Threads` and `Cpp11Threads`, where lower alignment could lead to false sharing, but these backends have no relevance in production code because of their poor performance anyway (and they could do not need to use the default anyway).

The performant CPU backends, should use no minimum alignment (=1) in favor of small allocations. Types with larger alignment requirements will be aligned appropriately anyway.

A small minimum alignment would however create a problem when allocating arrays of small types, where SIMD instructions might me generated with larger alignment. Maybe the alignment used for an allocation should not only be computed based on the types requirement but also take the size of the allocation. Maybe a template parameter `TDataMaxAlignBytes = core::vectorization::defaultAlignment` should be introduced to facilitate this.

**Comment by jkelling** [comment]
### Sub-issue
For `BlockSharedMemStMemberImpl`  the template param `TDataAlignBytes` was renamed to `TMinDataAlignBytes` in https://github.com/alpaka-group/alpaka/pull/1258/ , it apparently slipped by in this class, which should be fixed, next time is touched.

**Comment by psychocoderHPC** [comment]
Do I understand you correct that you would like to set the default alignment to `1` byte because each datum is aligned to `alignas(T)` anyway? 

IMO that's fine, the user can still enforce the alignment by align the data type.

**Comment by jkelling** [comment]
> Do I understand you correct that you would like to set the default alignment to `1` byte because each datum is aligned to `alignas(T)` anyway?

Yes. The default being not one is mostly a result of the "historical" origins of this class, where a minimum alignment of `4` is useful on GPU because of memory banks (i.e. we still need the template parameter for this purpose).

**Comment by jkelling** [comment]
> IMO that's fine, the user can still enforce the alignment by align the data type.

Yes. My only concern about this is, that this might soft-break existing code which allocates, e.g., an `array<int>` and does not force alignment to vector size, which is why I brought up arrays as potentially negatively affected by this. Although this is only a concern, if the old shared memory implementation of the CPU backends would have always aligned to `core::vectorization::defaultAlignment` instead of tho `alignof(T)`. Do you know which it was?

**Comment by psychocoderHPC** [comment]
You are right, we will possibly break the old code therefore we should not change the behavior for alpaka 0.6.1.
If we change this we should introduce it with alpaka 0.7.0
